zerver/worker/thumbnail.py

```diff
@@ -0,0 +1,133 @@
+import logging
+import time
+from dataclasses import asdict
+from io import BytesIO
+from typing import Any, Dict, Set
+
+import pyvips
+from django.db import transaction
+from typing_extensions import override
+
+from zerver.lib.mime_types import guess_type
+from zerver.lib.thumbnail import THUMBNAIL_OUTPUT_FORMATS, StoredThumbnailFormat
+from zerver.lib.upload import save_attachment_contents, upload_backend
+from zerver.models import ImageAttachment
+from zerver.worker.base import QueueProcessingWorker, assign_queue
+
+logger = logging.getLogger(__name__)
+
+
+@assign_queue("thumbnail")
+class ThumbnailWorker(QueueProcessingWorker):
+    @override
+    def consume(self, event: Dict[str, Any]) -> None:
+        start = time.time()
+        with transaction.atomic(savepoint=False):
+            try:
+                row = ImageAttachment.objects.select_for_update().get(id=event["id"])
+            except ImageAttachment.DoesNotExist:
+                logger.info("ImageAttachment row %d missing", event["id"])
+                return
+            uploaded_thumbnails = ensure_thumbnails(row)
+        end = time.time()
+        logger.info(
+            "Processed %d thumbnails (%dms)",
+            uploaded_thumbnails,
+            (end - start) * 1000,
+        )
+
+
+def ensure_thumbnails(image_attachment: ImageAttachment) -> int:
+    seen_thumbnails: Set[StoredThumbnailFormat] = set()
+    for existing_thumbnail in image_attachment.thumbnail_metadata:
+        seen_thumbnails.add(StoredThumbnailFormat(**existing_thumbnail))
+
+    needed_thumbnails = [
+        thumbnail_format
+        for thumbnail_format in THUMBNAIL_OUTPUT_FORMATS
+        if thumbnail_format not in seen_thumbnails
+    ]
```

Comment:
`THUMBNAIL_OUTPUT_FORMATS` seems to not be of type `StoredThumbnailFormat`, as noted in https://github.com/zulip/zulip/pull/30477/commits/cdeeb8eb4e2cc5f88532fc96e7a68cafff0da878#r1651723317; does that `in` check work?

Reply:
Yes.  These use the `__eq__` method from `BaseThumbnailFormat`, which only compare their stringifications, which are based on width, height, animation-ness, and extension.

Reply:
Ahh, OK. This seems fine, but we should make sure we have a test that verifies we didn't regress to "the seen_thumbnails check is always false".